[RPC] Create local RRef<ModuleInterface> remotely in Python, use it remotely in TorchScript

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34183 [RPC] Create local RRef<ModuleInterface> remotely in Python, use it remotely in TorchScript**
* #34292 [WIP] [RPC] Increase ref count before releasing GIL

https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to know the JIT type of it's contained `ScriptModule`.

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more straightforward to set it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to set an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

xush6528 committed:
[RPC] Create RRef<ModuleInterface> in Python, use it in TorchScript

`

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,46 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
+    return c10::nullopt;
+  }
+  c10::QualifiedName type_value_str = c10::QualifiedName(py::cast<std::string>(
+      py::module::import("torch.jit").attr("_qualified_name")(type_value)));
```
for non-existing type value, will this function return an empty option or throw an error?

xush6528 replied:
I think it should throw since it's a user provided type_hint, which should be mapped to a known JIT type, let me update.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -149,7 +149,10 @@ PyObject* rpc_init(PyObject* /* unused */) {
+          .def(
+              py::init<const py::object&, const py::object&>(),
+              py::arg("value"),
+              py::arg("type_hint") = py::none())
```
Do we want to expose API doc for this or this will this stay as an internal feature for now?

zzzwen replied:
We might no want to expose this `type_hint` field in Python.
It is possible that we handle it in c++?
The only known use case is to get around TorchScript limitation that most users should not care.

xush6528 replied:
Let's keep it private for now. 

This hidden default arg might be removed, since one alternative is to infer the type automatically for users. So, at this moment, I can't precisely foreseen what following RRef JIT typing changes we will make to it.

xush6528 replied:
@zzzwen This will bind to Python RRef.__init__.

mrshenli commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,46 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
+    return type_hint.value();
+  }
+
+  // Otherwise it's a spure pyobject, create the RRef
```
Just curious about the word spure. Looks like this is not english? Or do you mean pure?

xush6528 replied:
Good catch. I think it should be "pure pyobject".

mrshenli commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -101,10 +111,42 @@ def rref_script_annotation(rref_var):
+        self.assertEqual(ret, 10)
+
+        # Create a local RRef<MyModuleInterface>.
+        rref_script_module = rpc.RRef(MyScriptModule(3))
```
nit: in this case, `MyScriptModule`'s constructor arg is no longer `rank`.

xush6528 replied:
Let me update it to keep using self.rank.

mrshenli commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -101,10 +111,42 @@ def rref_script_annotation(rref_var):
+        rref = rpc.rpc_sync(dst_worker_name, owner_create_rref_my_script_module, args=(3,))
+        # Remote side uses the RRef in Script.
+        ret = rpc.rpc_sync(
+            rref.owner(), script_run_forward_rref_my_script_module, args=(rref,)
```
Will the error be propagated back properly if there is any?

xush6528 replied:
Added test for wrong type failure case.

xush6528 committed:
Update on "[RPC] Create RRef<ModuleInterface> in Python, use it in TorchScript"


`

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

zzzwen reviewed (commented):
Q: Can you add a test again "torchscript class interface"? not sure if this is the correct name, but the code is something like this

```
@torch.jit.interface
class Interface(object):
```

xush6528 committed:
Update on "[RPC] Create RRef<ModuleInterface> in Python, use it in TorchScript"


https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to track the type of the `ScriptModule` it holds..

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more helpful to mark it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to mark an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

xush6528 commented:
test_torchscript_function is failing. need to investigate.

xush6528 committed:
Update on "[RPC] Create RRef<ModuleInterface> locally in Python, use it remotely in TorchScript"


https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to know the JIT type of it's contained `ScriptModule`.

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more straightforward to set it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to set an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -113,6 +145,61 @@ def rref_script_annotation(rref_var):
+
+        # Use RRef<MyModuleInterface> remotely in Script with wrong argument type hint.
+        fut = rpc.rpc_async(
+            rref.owner(), script_run_forward_rref_my_script_module_2, args=(rref,)
```
Just want to check if I understand this correctly. This fails because it expects a `MyModuleInterface2`, but the RRef actually holds a `MyModuleInterface`?

xush6528 replied:
Yes.

wanchaol replied:
why we need to do this check? I think this is testing the jit function type matching for module interface which already been covered somewhere else.

xush6528 replied:
@wanchaol Good point. I think it also add burdens if one day JIT framework decide to throw a different error message. They would be astonished to see this RPC test is broken. Let me remove it.

mrshenli reviewed (approved):
LGTM! Thanks for adding this!

wanchaol commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,55 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
+  if (module.has_value()) {
+    TORCH_CHECK(
+        type_hint_ptr != nullptr,
+        "The RRef being created contains a ScriptModule, must provide it's ModuleInterface type hint.");
```
nit: its

xush6528 replied:
Updated.

wanchaol commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,55 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
+    TORCH_CHECK(
+        type_hint_ptr != nullptr,
+        "The RRef being created contains a ScriptModule, must provide it's ModuleInterface type hint.");
+    TORCH_CHECK(
```
can we make these two checks into one? 
```
type_hint_ptr != nullptr && type_hint_ptr->kind() == TypeKind::InterfaceType

TORCH_CHECK will exist in release build and we should reduce its usage if possible

xush6528 replied:
Merged.

wanchaol commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,55 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
+    TORCH_CHECK(
+        type_hint_ptr->kind() == TypeKind::InterfaceType,
+        "The RRef being created contains a ScriptModule, must provide it's ModuleInterface type hint");
+    TORCH_CHECK(
```
considering merging this together too.

xush6528 replied:
Merged.

wanchaol commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -54,18 +104,17 @@ PyRRef::PyRRef(c10::intrusive_ptr<RRef> rref) : rref_(std::move(rref)) {
+              elem_type->isSubtypeOf(type_hint_ptr),
+              "The specificied type hint is ",
+              type_hint_ptr->python_str(),
+              "The decided type is ",
```
that's kind of confusing to user, user shouldn't know what a specified type or a decided type means. We should just error out and saying: 
type_hint_ptr->python_str() is not a subtype of elem_type->python_str()

wanchaol replied:
also, this seems to be already checked in `decidePyObjJitType` from what i see. 

For the API usage of this `type_hint`, We should only allow user to specify the type_hint when the type_hint is a jit class type/interface type. That means if user pass a type that is not a class type, we should guard against it, because that's also what `tryTypePyObjToTypePtr` is trying to do. But the current state of this PR allows user to possibly pass types other than class types (like Int, Tensor, etc.) and will errors with `Please script the type to register it before using it.` we should instead tell the user sth like "type_hint only accepts TorchScript class/interface"

xush6528 replied:
Right, I am allowing any type hint. Let me narrow it down to ScriptModule. 
Need to add test case for ScriptClass.

wanchaol commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,55 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
+      type_ptr != nullptr,
+      "Type, ",
+      type_qualified_name.qualifiedName(),
+      ", has not been registered in JIT CompilationUnit yet. "
```
please don't throw this error message, `JIT CompilationUnit` is not something the user should know about. We should just say: 

`type_hint` could not be found, did you pass a valid interface type?

xush6528 replied:
Good suggestion. Updated.

wanchaol reviewed (changes requested):
There're several type check/ user facing error message we should fix before we merge this.

wanchaol commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,55 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
   return RRefForkData(ownerId, rrefId, forkId, parent, typeStr);
 }
+
+TypePtr tryTypePyObjToTypePtr(const py::object& type_value) {
```
I am thinking we should just merge `tryTypePyObjToTypePtr` with `decidePyObjJitType`, these two functions are quite confusing and take a while to know what they are really doing.

Maybe something like `tryToInferTypeFromPyObject(const py::object& value, const py::object& type_hint)`

xush6528 replied:
Merged.

xush6528 committed:
Update on "[RPC] Create RRef<ModuleInterface> locally in Python, use it remotely in TorchScript"


https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to know the JIT type of it's contained `ScriptModule`.

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more straightforward to set it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to set an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

xush6528 committed:
Update on "[RPC] Create RRef<ModuleInterface> locally in Python, use it remotely in TorchScript"


https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to know the JIT type of it's contained `ScriptModule`.

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more straightforward to set it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to set an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

xush6528 committed:
Update on "[RPC] Create RRef<ModuleInterface> locally in Python, use it remotely in TorchScript"


https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to know the JIT type of it's contained `ScriptModule`.

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more straightforward to set it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to set an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

wanchaol reviewed (commented):
looking good, some nits and some questions.

xush6528 commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -113,6 +155,77 @@ def rref_script_annotation(rref_var):
+class LocalRRefTest(RpcAgentTestFixture):
+    @dist_init
+    def test_create_local_script_class_rref_in_py(self):
+        if self.rank != 0:
```
This test method is run by every process, which is a useful behavior for data-parallel tests. 
bBut for RPC stuff, it's not very useful. S

The behavior is controlled by the code around this PR, https://github.com/pytorch/pytorch/pull/34150.

I want to spend time to get rid of this behavior for RPC tests. But for now, I make other processes idle by this trick.

I will find time to see if we can let only process run this test method.

xush6528 commented on torch/testing/_internal/distributed/rpc/jit/rpc_test.py:
```diff
@@ -48,8 +48,17 @@ def forward(self):
+        # type: (int)
+        self.a = a
+
+    def increment_value(self, increment):
```
Removing for now.

xush6528 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,57 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
+  // If the py::object to be contained by the RRef is a ScripModule, we enforce
+  // users to specify its ModuleInterface type.
+  c10::optional<jit::script::Module> module(jit::script::as_module(value));
+  if (module.has_value()) {
```
this is neat!

xush6528 commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -46,6 +47,57 @@ RRefForkData fromPyTuple(const py::tuple& pyTuple) {
+  c10::optional<jit::script::Module> module(jit::script::as_module(value));
+  if (module.has_value()) {
+    TypePtr type_hint_ptr;
+    if (!type_hint.is(py::none())) {
```
Sounds better. But that will be multiple TORCH_CHECK, since they can only be checked step by step.

xush6528 committed:
Update on "[RPC] Create RRef<ModuleInterface> locally in Python, use it remotely in TorchScript"


https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to know the JIT type of it's contained `ScriptModule`.

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more straightforward to set it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to set an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/rpc/py_rref.cpp:
```diff
@@ -53,34 +53,30 @@ TypePtr tryInferTypeWithTypeHint(
+                                  .attr("_qualified_name")(type_hint)));
+    TypePtr type_hint_ptr =
+        jit::get_python_cu()->get_interface(type_qualified_name);
+    TORCH_CHECK(
```
We could do
```
TORCH_CHECK(
type_hint_ptr != nullptr && module.value().type()->isSubtypeOf(type_hint_ptr),
module.value().type()->python_str(),
" is not a subtype of the type hint: ",
type_qualified_name.qualifiedName(),
", did you pass a valid interface type?");
```

class_type->python_str() just return its underlying qualified name I think.

xush6528 replied:
Right. This is nice!

wanchaol reviewed (approved):
looks good! one minor suggestion on merging two checks.

xush6528 committed:
Update on "[RPC] Create RRef<ModuleInterface> locally in Python, use it remotely in TorchScript"


https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to know the JIT type of it's contained `ScriptModule`.

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more straightforward to set it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to set an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]

xush6528 committed:
Update on "[RPC] Create local RRef<ModuleInterface> remotely in Python, use it remotely in TorchScript"


https://github.com/pytorch/pytorch/pull/33263 enhanced the RRef Python constructor to infer most types, by `jit::tryToInferType(..)`. 

But this helper function can't infer `ScriptModule` type due to `ScriptModule`'s special per-Module type singleton logic, so it's still not possible for an Python-created RRef to know the JIT type of it's contained `ScriptModule`.

Instead of inferring the specific type of a Module, which could leads to too many candidate types (due to Module's multiple inheritance possibility), it's more straightforward to set it's type as a user-specified `ModuleInterface` type.

We added an optional argument `type_hint` for users to set an `RRef` for what `ModuleInterface` type it's holds.

Differential Revision: [D7065050](https://our.internmc.facebook.com/intern/diff/D7065050/)

[ghstack-poisoned]
